### PR TITLE
Imap capability implementation.

### PR DESCRIPTION
**Pull request by kajojify**
#13

### CODE REVIEW

File: heralding/capabilities/imap.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            return "Not Authenticated"
+
+        if auth_mechanism in self.available_mechanisms:
+            # the space after '+' is needed according to RFC
+            self.send_message(session, gsocket, '+ ')
+            raw_msg = gsocket.recv(512)
+
+            if auth_mechanism == 'plain':
+                success, credentials = self.try_b64decode(raw_msg, session)
+                if success and credentials.count('\x00') == 2:
```
Comment (johnnykv) [inline heralding/capabilities/imap.py]: a comment here explaining the \00 would be nice.

File: heralding/capabilities/imap.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                    _, user, password = base64.b64decode(raw_msg).split('\x00')
+                    session.add_auth_attempt('plaintext', username=user, password=password)
+                    self.send_message(session, gsocket, tag + ' NO Authentication failed')
+                else:
+                    self.send_message(session, gsocket, tag + ' BAD invalid command')
+        else:
+            self.send_message(session, gsocket, tag + ' BAD invalid command')
+        self.stop_if_too_many_attempts(session)
+        return 'Not Authenticated'
```
Comment (johnnykv) [inline heralding/capabilities/imap.py]: Again, do we use this return value anywhere?

File: heralding/capabilities/imap.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
+# Copyright (C) 2017 Johnny Vestergaard <jkv@unixcluster.dk>
```
Comment (kajojify) [inline heralding/capabilities/imap.py]: I tried to follow Heralding general style of coding, so I based Imap class on Pop3 class. 
Thus I think, it is still yours copyright.

File: heralding/capabilities/imap.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                    self.send_message(session, gsocket, tag + " BAD invalid command")
+        session.end_session()
+
+    def cmd_authenticate(self, session, gsocket, tag, args):
+        mechanism = args.split()
+        if len(mechanism) == 1:
+            auth_mechanism = mechanism[0].lower()
+        else:
+            self.send_message(session, gsocket, tag + ' BAD invalid command')
+            return "Not Authenticated"
```
Comment (kajojify) [inline heralding/capabilities/imap.py]: "Not Authenticated" value is used in comparison with "Logout" in while loop:

```
while state != 'Logout' and session.connected:
     try:
         raw_msg = fileobj.readline()
         except socket.error:
```
I decided that each method should return the string, which means current state of Imap session. It isn't important, but in my opinion it is simpler to understand, how each Imap command affects on the session state, while looking through the code.

File: heralding/capabilities/imap.py (hunk 1 above)
Comment (johnnykv) [inline heralding/capabilities/imap.py]: Okay, will merge as soon when you comment that line.

**Comment by johnnykv** [comment]
close #13

**Pull request merged**